### PR TITLE
nixos: disable sound by default, if stateVersion >= 18.03

### PR DESCRIPTION
**Pull request by aristidb**
###### Motivation for this change

This change was suggested by @fpletz in #35292. Sound is disabled by default (for stateVersion > 18.03), and nixos-generate-config now puts a helpful comment telling you how to enable it. This is consistent with X11 being disabled by default.

###### Things done

I'm still building the branch, so not ready yet with everything. I wanted to create the PR already to enable discussion.

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by aristidb** [comment]
CC @fpletz @vcunat @grahamc

File: nixos/modules/services/audio/alsa.nix [hunk 1]
```diff
@@ -78,7 +77,9 @@ in
 
   ###### implementation
 
-  config = mkIf config.sound.enable {
+  config = {
+      sound.enable = mkDefault !(versionAtLeast config.system.stateVersion "18.03")
```
Comment (grahamc) [inline nixos/modules/services/audio/alsa.nix]: you need a `;` at the end of this line

File: nixos/modules/services/audio/alsa.nix [hunk 2]
```diff
@@ -78,7 +77,9 @@ in
...
+    } // mkIf config.sound.enable {
```
Comment (grahamc) [inline nixos/modules/services/audio/alsa.nix]: I'm not sure, but you _might_ need to wrap the `mkIf config.sound.enable { ... }` in `(...)`.

**Comment by grahamc** [comment]
Try testing it like this: `HOME=/homeless-shelter NIX_PATH=nixpkgs=$(pwd) nix-instantiate ./nixos/release.nix -A options --option restrict-eval true --option build-timeout 1800 --argstr system x86_64-linux --show-trace`

**Comment by oxij** [comment]
Agree in principle, strongly against in practice:

- Breaks existing configurations.
- Headless and speakerless systems disable it anyway.
- Saves only about 2.3Mb (can be made ~828K if we separate outputs of `alsa-utils`) anyway.

**Comment by grahamc** [comment]
How does this break existing configurations? Are people deleting the `do-not-touch` `stateVersion` from their configuration? If so, I'm okay with that and we should move forward.

**Comment by oxij** [comment]
Also, what is #35319?

**Comment by aristidb** [comment]
@oxij Thanks for point out the other PR that I missed.

**Comment by oxij** [comment]
> How does this break existing configurations? Are people deleting the `do-not-touch` `stateVersion` from their configuration? If so, I'm okay with that and we should move forward.

I use and continuously update the same `configuration.nix` from around 7 years ago (and "reinstall" by simply copying system closure to a new HDD). My configuration does not contain such a line.

**Comment by grahamc** [comment]
I understand. I'm okay with this level of breakage, as the stateVersion has been around for several years now, is well documented, and long time users have had several opportunities (and will have another opportunity) to catch this problem via the release notes before their configuration breaks.

**Comment by oxij** [comment]
> I understand. I'm okay with this level of breakage, as the stateVersion has been around for several years now, is well documented, and long time users have had several opportunities (and will have another opportunity) to catch this problem via the release notes before their configuration breaks.

Ugh, can't we at the very least generate a warning if you don't have `stateVersion` set? Surely I'm not alone here.

Now, as to the implementation.

```
+  # Enable sound.
+  # sound.enable = true;
+  # hardware.pulseaudio.enable = true;
+
```

`hardware.pulseaudio.enable = true;` line is out of scope of this PR and belongs to the discussion of #35292. Lets not repeat all of that here. "Hey, its just a comment" is not an argument. I'm absolutely serious. Please, remove that line.

```
-  config = mkIf config.sound.enable {
+  config = {
+      sound.enable = mkDefault (!versionAtLeast config.system.stateVersion "18.03");
+    } // mkIf config.sound.enable {
```

This is ugly and probably won't work (lacks parens). Why don't you just put that expression into the `default = ` where it belongs? The manual build can then be made deterministic again by defining `defaultText`.

**Comment by aristidb** [comment]
@oxij Your aggressive style is not appreciated.

**Comment by oxij** [comment]
How is any of the above aggressive?

**Comment by aristidb** [comment]
I've fixed the error of how I was using mkIf in 473cfed (my nixos module system skills are rusty :(), and the missing defaultText in f372898. Fortunately, this will all be squash-merged, so nobody will have to look at my erroneous intermediate commits.

Now I'm waiting for @fpletz and @vcunat to say what they think about getting this into 18.03.

**Comment by oxij** [comment]
... and then most of my substantial comments were ignored =/

See https://github.com/oxij/nixpkgs/commit/42a814ee2648f0eb9e5169a7f648c738d99ca921 instead.

**Comment by oxij** [comment]
Same thing, but with warnings and no sneaky changes: #35366.

**Comment by grahamc** [comment]
> How is any of the above aggressive?

Jan,

Let me start with recognizing that you don't care for PulseAudio and
possibly the entire FreeDesktop software universe. You have made your
opinion of Poettering quite clear. I appreciate your position on
this, and respect your choice.

As an organizational goal, the NixOS system defaults should do a nice
job representing a common, well supported base system to work off of.
It is especially nice when this default also matches what most users
are already using, and what upstream software writers are expecting.

In the case of Firefox, indeed -- upstream expects PulseAudio and
we're inclined to follow that lead. NixOS does not have the user base
or contributor base to put up a meaningful resistance to far-reaching
ecosystem changes like this.

In the case of users, because NixOS has no sort of metrics collection
mechanism, it is much harder to tell. However, anecdotally, it is
a very common option to enable -- present in most public
configurations and tutorials I've come across.

It seems to be a very common opinion that NixOS should in fact enable
PulseAudio if sound is enabled by default, and indeed I also support
that position. 

However, my position on PulseAudio is not what I'm writing this for,
it is more about your approach to communicating about the issue. Your
behavior on these recent pull requests have distressed, upset, and
frustrated fellow members of the community.

Firstly: it is not appropriate to compare a configuration setting
about sound to the Stasi, or a re-framing of a poem about the
systematic extermination of people.

Secondly: a much easier way to explain the PRs around enabling
PulseAudio is that it is considered an improvement to the defaults.
Not an attack upon you, not conspiracy against you.

People have different opinions from you, and that is okay. This is a
beautiful part of the NixOS configuration mechanism!

When people do things you disagree with, it is not stupid. When people
do things you disagree with, it is not crazy. It is not appropriate to
question a fellow contributor's mental fitness because you don't care
for the work they've done.

Franz is not lying when he interprets his experience and observations
and concludes that most NixOS users are using PulseAudio. It is not
appropriate to accuse well-meaning core contributors of lying.

Finally, nobody is forcing you to use any software. You can choose to
not use NixOS. In a lot of ways, you already have -- by making SLNOS.
NixOS has no intentions of making SLNOS hard to maintain, and at the
same time NixOS is not beholden to making it easy or convenient for
SLNOS to exist.

If you truly believe NixOS contributors are crazy, lying, and
conspiring against you, then I wholeheartedly recommend you avoid
such a toxic environment. Otherwise, please tone down your rhetoric
and personal attacks. 
 
Thank you,
Graham

**Comment by grahamc** [comment]
> How is any of the above aggressive?

Jan,

Let me start with recognizing that you don't care for PulseAudio and
possibly the entire FreeDesktop software universe. You have made your
opinion of Poettering quite clear. I appreciate your position on
this, and respect your choice.

As an organizational goal, the NixOS system defaults should do a nice
job representing a common, well supported base system to work off of.
It is especially nice when this default also matches what most users
are already using, and what upstream software writers are expecting.

In the case of Firefox, indeed -- upstream expects PulseAudio and
we're inclined to follow that lead. NixOS does not have the user base
or contributor base to put up a meaningful resistance to far-reaching
ecosystem changes like this.

In the case of users, because NixOS has no sort of metrics collection
mechanism, it is much harder to tell. However, anecdotally, it is
a very common option to enable -- present in most public
configurations and tutorials I've come across.

It seems to be a very common opinion that NixOS should in fact enable
PulseAudio if sound is enabled by default, and indeed I also support
that position. 

However, my position on PulseAudio is not what I'm writing this for,
it is more about your approach to communicating about the issue. Your
behavior on these recent pull requests have distressed, upset, and
frustrated fellow members of the community.

Firstly: it is not appropriate to compare a configuration setting
about sound to the Stasi, or a re-framing of a poem about the
systematic extermination of people.

Secondly: a much easier way to explain the PRs around enabling
PulseAudio is that it is considered an improvement to the defaults.
Not an attack upon you, not conspiracy against you.

People have different opinions from you, and that is okay. This is a
beautiful part of the NixOS configuration mechanism!

When people do things you disagree with, it is not stupid. When people
do things you disagree with, it is not crazy. It is not appropriate to
question a fellow contributor's mental fitness because you don't care
for the work they've done.

Franz is not lying when he interprets his experience and observations
and concludes that most NixOS users are using PulseAudio. It is not
appropriate to accuse well-meaning core contributors of lying.

Finally, nobody is forcing you to use any software. You can choose to
not use NixOS. In a lot of ways, you already have -- by making SLNOS.
NixOS has no intentions of making SLNOS hard to maintain, and at the
same time NixOS is not beholden to making it easy or convenient for
SLNOS to exist.

If you truly believe NixOS contributors are crazy, lying, and
conspiring against you, then I wholeheartedly recommend you avoid
such a toxic environment. Otherwise, please tone down your rhetoric
and cease with the personal attacks. 
 
Thank you,
Graham

**Comment by edolstra** [comment]
Hm, this is not really the intended use of `stateVersion`. From the description:
```
        Every once in a while, a new NixOS release may change
        configuration defaults in a way incompatible with stateful
        data. For instance, if the default version of PostgreSQL
        changes, the new version will probably be unable to read your
        existing databases. To prevent such breakage, you can set the
        value of this option to the NixOS release with which you want
        to be compatible. The effect is that NixOS will option
        defaults corresponding to the specified release (such as using
        an older version of PostgreSQL).
```
So this is only intended for options that have some corresponding on-disk state. AFAICT this is not the case for sound. In any case `stateVersion` is a necessary evil that only exists because we can't just upgrade Postgres databases or change SSH host keys. It's not necessary for things like whether sound is enabled. (If the user discovers that sound is suddenly disabled, they can just enable it.)

I had some vague recollection that we also had a `configVersion` option setting to control the defaults for non-state-related options, but I can't find it so maybe it was only discussed.

File: nixos/modules/installer/tools/nixos-generate-config.pl [hunk 3]
```diff
@@ -603,6 +603,10 @@ sub multiLineList {
   # Enable CUPS to print documents.
   # services.printing.enable = true;
 
+  # Enable sound.
+  # sound.enable = true;
+  # hardware.pulseaudio.enable = true;
```
Comment (edolstra) [inline nixos/modules/installer/tools/nixos-generate-config.pl]: BTW, shouldn't `hardware.pulseaudio.enable = true` imply `sound.enable = true`? Or perhaps the other way around, if we want Pulseaudio to be the default sound system.

**Comment by oxij** [comment]
> BTW, shouldn't `hardware.pulseaudio.enable = true` imply `sound.enable = true`? Or perhaps the other way around, if we want Pulseaudio to be the default sound system.

They are independent.

`sound.enable = true` makes `asound.conf` and runs ALSA services, `hardware.pulseaudio.enable = true` makes its own `asound.conf` with ALSA->PA adapter and runs PA services.

**Comment by vcunat** [comment]
Maybe the defaults would better be solved along these lines:
```nix
{
  # X, pulseaudio, ...
  imports = [ <nixpkgs/nixos/modules/profiles/desktop.nix> ];
}
```

**Comment by oxij** [comment]
Probably a good idea to point to this https://github.com/NixOS/nixpkgs/pull/35292#issuecomment-368217334 here too.

**Comment by abbradar** [comment]
Linking https://github.com/NixOS/nixpkgs/commit/e349ccc77febd45abbd14be14f7de123ec4a4da2#commitcomment-27757112 here as I'm fairly concerned about making this change unconditionally -- it would break every NixOS desktop's user sound after an update, which IMO is not justified by a small closure size benefit in default server setups.

**Comment by abbradar** [comment]
To be clear I was okay with this change as long as it was conditional on some variable which specifies how "modern" is current NixOS configuration. However I also agree with @edolstra that `stateVersion` is unfit for this so we have no good way of determining.

Offtopic: @oxij made a good point that we should warn if `stateVersion` is not set. Instead we should have a special value that says "I want bleeding edge state assumptions and don't mind consequences".

**Comment by oxij** [comment]
> Offtopic: @oxij made a good point that we should warn if `stateVersion` is not set.

Did you look at #35366?

**Comment by abbradar** [comment]
@oxij Yeah, I actually wanted to propose you to split `stateVersion` changes into a separate PR.

**Comment by Mic92** [comment]
This pull request broke nixops

**Comment by oxij** [comment]
Let me remind you that after this was reverted the ALSA module is now broken by default. If you don't want to revert the original change, let's make a warning at the very least. See #46490.

**Pull request merged**